sklearn/compose/_column_transformer.py

```diff
@@ -0,0 +1,616 @@
+        get_weight = (self.transformer_weights or {}).get
+
+        for name, trans, column in transformers:
+            if X is None:
+                sub = X
+            else:
+                sub = _get_column(X, column)
+
+            if replace_strings:
+                # replace 'passthrough' with identity transformer and
+                # skip in case of 'drop'
+                if trans == 'passthrough':
+                    trans = FunctionTransformer(
+                        validate=False, accept_sparse=True,
+                        check_inverse=False)
+                elif trans == 'drop':
+                    continue
+
+            yield (name, trans, sub, get_weight(name))
+
+    def _validate_transformers(self):
+        names, transformers, _, _ = zip(*self._iter())
+
+        # validate names
+        self._validate_names(names)
+
+        # validate estimators
+        for t in transformers:
+            if t in ('drop', 'passthrough'):
+                continue
+            if (not (hasattr(t, "fit") or hasattr(t, "fit_transform")) or not
+                    hasattr(t, "transform")):
+                raise TypeError("All estimators should implement fit and "
+                                "transform, or can be 'drop' or 'passthrough' "
+                                "specifiers. '%s' (type %s) doesn't." %
+                                (t, type(t)))
+
+    def _validate_remainder(self, X):
+        """Generate list of passthrough columns for 'remainder' case."""
+        if self.remainder not in ('drop', 'passthrough'):
+            raise ValueError(
+                "The remainder keyword needs to be one of 'drop' or "
+                "'passthrough'. {0:r} was passed instead")
+
+        n_columns = X.shape[1]
+
+        if self.remainder == 'passthrough':
+            cols = []
+            for _, _, columns in self.transformers:
+                cols.extend(_get_column_indices(X, columns))
+            self._passthrough = sorted(list(set(range(n_columns)) - set(cols)))
+            if not self._passthrough:
+                # empty list -> no need to select passthrough columns
+                self._passthrough = None
+        else:
+            self._passthrough = None
+
+    @property
+    def named_transformers_(self):
+        """Access the fitted transformer by name.
+
+        Read-only attribute to access any transformer by given name.
+        Keys are transformer names and values are the fitted transformer
+        objects.
+
+        """
+        # Use Bunch object to improve autocomplete
+        return Bunch(**dict([(name, trans) for name, trans, _
+                             in self.transformers_]))
+
+    def get_feature_names(self):
+        """Get feature names from all transformers.
+
+        Returns
+        -------
+        feature_names : list of strings
+            Names of the features produced by transform.
+        """
+        check_is_fitted(self, 'transformers_')
```

Comment:
Shouldn't remainder be handled here too?

Reply:
Ideally, yes. I am only not fully sure what to do here currently, given that `get_feature_names` is in general not really well supported. 
I think ideally I would add the names of the passed through columns to `feature_names`, but then the actual string column names in case of pandas dataframes. And in case of numpy arrays return the indices into that array as strings? (['0', '1', ..])

I can also raise an error for now if there are columns passed through, just to make sure that if we improve the `get_feature_names` in the future, it does not lead to a change in behaviour (but a removal of the error).

Reply:
Can just raise NotImplementedError in case of `remainder != 'drop'` for now.... Or you can tack the remainder transformer onto the end of `_iter`.

I agree  `get_feature_names` is not quite the right design.

Reply:
Can we have an issue for this?